### ISSUE TITLE
ConcurrentModificationException on read

### ISSUE THREAD

**Issue by sghpjuikit** [haraldk/TwelveMonkeys#402]
Reading an image in concurrent environment fails. I'm reading images in a threadpool with size up to 8 threads. The issue was observed on build from latest sources on master branch.

This particular error happened when reading a thumbnail, which I do when the size is greater than requested size, but I think it appears when reading the image itself as well.

````
java.util.ConcurrentModificationException: null
	at java.base/java.util.Vector$Itr.checkForComodification(Vector.java:1295) ~[na:na]
	at java.base/java.util.Vector$Itr.next(Vector.java:1251) ~[na:na]
	at java.desktop/sun.java2d.cmm.ProfileDeferralMgr.activateProfiles(ProfileDeferralMgr.java:93) ~[na:na]
	at java.desktop/java.awt.color.ICC_Profile.getInstance(ICC_Profile.java:788) ~[na:na]
	at java.desktop/com.sun.imageio.plugins.jpeg.JPEGImageReader.setImageData(JPEGImageReader.java:771) ~[na:na]
	at java.desktop/com.sun.imageio.plugins.jpeg.JPEGImageReader.readImageHeader(Native Method) ~[na:na]
	at java.desktop/com.sun.imageio.plugins.jpeg.JPEGImageReader.readNativeHeader(JPEGImageReader.java:723) ~[na:na]
	at java.desktop/com.sun.imageio.plugins.jpeg.JPEGImageReader.checkTablesOnly(JPEGImageReader.java:347) ~[na:na]
	at java.desktop/com.sun.imageio.plugins.jpeg.JPEGImageReader.gotoImage(JPEGImageReader.java:493) ~[na:na]
	at java.desktop/com.sun.imageio.plugins.jpeg.JPEGImageReader.readHeader(JPEGImageReader.java:716) ~[na:na]
	at java.desktop/com.sun.imageio.plugins.jpeg.JPEGImageReader.readInternal(JPEGImageReader.java:1173) ~[na:na]
	at java.desktop/com.sun.imageio.plugins.jpeg.JPEGImageReader.read(JPEGImageReader.java:1153) ~[na:na]
	at java.desktop/javax.imageio.ImageReader.read(ImageReader.java:938) ~[na:na]
	at com.twelvemonkeys.imageio.plugins.jpeg.ThumbnailReader.readJPEGThumbnail(ThumbnailReader.java:72) ~[imageio-jpeg-3.3.9.jar:3.3.9]
	at com.twelvemonkeys.imageio.plugins.jpeg.EXIFThumbnailReader.readJPEG(EXIFThumbnailReader.java:138) ~[imageio-jpeg-3.3.9.jar:3.3.9]
	at com.twelvemonkeys.imageio.plugins.jpeg.EXIFThumbnailReader.readJPEGCached(EXIFThumbnailReader.java:103) ~[imageio-jpeg-3.3.9.jar:3.3.9]
	at com.twelvemonkeys.imageio.plugins.jpeg.EXIFThumbnailReader.getWidth(EXIFThumbnailReader.java:221) ~[imageio-jpeg-3.3.9.jar:3.3.9]
	at com.twelvemonkeys.imageio.plugins.jpeg.JPEGImageReader.getThumbnailWidth(JPEGImageReader.java:1095) ~[imageio-jpeg-3.3.9.jar:3.3.9]````

**Comment by haraldk** [comment]
Hi Martin,

Please not that the ImageIO API specifically states that an `ImageReader` instance is not required to be thread safe. In particular, the API design is such that it's not convenient to even try to make them thread safe. However, as long as you don't share instances of `ImageReader`s (or `ImageInputStream`s) between threads, you should be good.

However, the exception you experience above, seems to be a JDK issue, and must be reported to Oracle or the OpenJDK community.

The method [`sun.java2d.cmm.ProfileDeferralMgr.activateProfiles()`](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/sun/java2d/cmm/ProfileDeferralMgr.java#ProfileDeferralMgr.activateProfiles%28%29)  internally loops over a `Vector`, then removes all elements in that same vector. If it invoked twice, the second attempt may throw exception from the loop, as the vector is concurrently modified (`removeAllElements()`). 

There's nothing we can reasonably do to prevent this. You *may* be able work around the issue by loading an ICC profile in a static block in your code (ie. `ICCProfile.getInstance(ColorSpace.CS_sRGB)`) before attempting to read an image.


-- 
Harald K

**Comment by sghpjuikit** [comment]
I see.

I overlooked the fact that activateProfiles() is JDK and this appears to be an old [old bug](https://bugs.openjdk.java.net/browse/JDK-6986863), unlikely to be fixed. That is disheartening.  I see you were [aware](https://stackoverflow.com/questions/26297491/imageio-thread-safety) already.

As for the workaround, I'm skeptical. What leads you to believe ICCProfile can be _loaded_?
I'm not familiar with color profiles. How would I determine which profile to load?

I don't know, maybe we can close this? I'll investigate some more and comment if I come up with something.

**Comment by haraldk** [comment]
Sorry, forgot I had seen this before, but yes, that is the same issue.

If you are on a different platform than Linux/Ubuntu, you might want to report the issue again, to give it more attention. The fix for this seems pretty trivial if you ask me. Either making the method synchronized or use some other means of guards/locks inside the method.

About the workaround, I'm sorry, I looked at the code again, and just using what I wrote above will *not* load (activate) the profile. It will simply return a (possibly still "deferred") profile.

However, the following will (`getData()` method will force activation of all deferred profiles):

````java
static {
    // Will in turn force invocation to ProfileDeferralMgr.activateProfiles()
    ICC_Profile.getInstance(ColorSpace.CS_sRGB).getData();
}
````

Doing this in a static block, will make sure it happens once and only once, during class loading (initialization). This is guarded by the JLS, so we can safely count on it.

Closing this for now.


Best regards,

-- 
Harald K

**Comment by sghpjuikit** [comment]
Thank you for the additional feedback. I do have to ask though:
- How do you know that once you activate the profiles, it will never happen again (thus solving the issue)?
- How do you know this will load all required profiles?

I'm definitely giving this a try, but I've had trouble reproducing the error so far.

**Comment by haraldk** [comment]
Hi again,

> How do you know that once you activate the profiles, it will never happen again (thus solving the issue)?

This is my understanding, based on examining the JDK source code. Of course, using some other JDK than the Oracle JDK or OpenJDK-based JDK, this will likely not have any effect (but the bug might not be there in the first place). Relying on implementation details like this is never optimal or bullet proof (especially not future proof). But in cases of long-standing bugs like this, it might be the best option we have.

You will see that all invocations of `activateProfiles()` is guarded by an `if (deferring)`, like this:

````java
    if (ProfileDeferralMgr.deferring) {
        ProfileDeferralMgr.activateProfiles();
    }
````

For example in `ICC_Profile.getData()` or `ColorConvertOp` static initializer block.

See [`ProfileDeferralMgr.activateProfiles()` method](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/sun/java2d/cmm/ProfileDeferralMgr.java#ProfileDeferralMgr.activateProfiles%28%29), and you'll see that the first thing the method does, is setting the `deferring` class variable to `false`, which should prevent the method from even being invoked again. 

But of course, it doesn't, as we have seen, as the new value isn't immediately propagated to all threads, so simultaneous threads may still hold the old cached `true` value. But if we can invoke this method early on, before setting up the thread pool etc, we should effectively avoid (or at the very least, minimize the chance of) multiple simultaneous invocations.

> How do you know this will load all required profiles?

It won't. However, invoking this method will *activate* all *deferred* profiles, and disable further deferral. 

Again, I see this by reading the source code. The attempted activation of all deferred profiles is what is causing the exception in the first place, due to the above mentioned lack of synchronization. 

You'll also see that, once the `deferring` flag has been set to `false`, no further attempts to defer profiles is made, and the flag is never set back to `true`.


Best regards,

-- 
Harald K

**Comment by sghpjuikit** [comment]
I see.
Thank you for clear and very helpful explanation.

**Comment by sghpjuikit** [comment]
Just an update, I still was not able to reproduce the issue, hence your workaround may have worked.

**Comment by haraldk** [comment]
Thanks for reporting back and providing feedback! I might suggest this as a workaround in the readme in the future.

-- 
Harald K

**Comment by sghpjuikit** [comment]
I'm back. After period of use, I have seen no ConcurrentModificationException. It's not 100%, but I consider the issue resolved and the solution satisfactory. I'm posting the entirety of the code that I've been using.

`        // Load deferred color space profiles to avoid ConcurrentModificationException due to JDK
        // Use in public static main void or prior to application initialization
        // https://github.com/haraldk/TwelveMonkeys/issues/402
        // https://bugs.openjdk.java.net/browse/JDK-6986863
        // https://stackoverflow.com/questions/26297491/imageio-thread-safety
        ICC_Profile.getInstance(ColorSpace.CS_sRGB).getData();
        ICC_Profile.getInstance(ColorSpace.CS_PYCC).getData();
        ICC_Profile.getInstance(ColorSpace.CS_GRAY).getData();
        ICC_Profile.getInstance(ColorSpace.CS_CIEXYZ).getData();
        ICC_Profile.getInstance(ColorSpace.CS_LINEAR_RGB).getData();`

Thx for the help.

**Comment by mrserb** [comment]
In JDK17 the deferral mechanic was reworked and this particular exception was fixed, please take a look and report any issues you may found.

**Comment by haraldk** [comment]
Thanks @mrserb,

I had a brief look at the commit in OpenJDK and it looked reasonable. 😉 

Thanks a lot for fixing this bug! 

PS: Do you know if the fix will be backported to earlier Java versions (e.g 11 LTS)? If so, we could perhaps remove the `ICC_Profile.getInstance(...).getData()` hack we're now using in the library. Then again, it probably won't do much harm either...

Thanks,

-- 
Harald K
